doc: update contributing guidelines and add issue/PR templates

nh13 opened the pull request:
Based on the rust-by-example contributing doc: https://github.com/rust-lang/rust-by-example

Now includes:

- Code of Conduct - Basic statement about maintaining a welcoming environment
- Getting Started - Prerequisites and step-by-step development environment setup
- Development Workflow - Branch management, making changes, and pre-commit checks
- Issue Tracker - Guidelines for reporting bugs, suggesting features, and common labels
- Pull Requests - Pre-submission checklist, submission process, and review expectations
- Coding Guidelines - Style (rustfmt, clippy), best practices, and dependency policy
- Testing - How to run tests and guidelines for writing new tests
- Documentation - Kept the original README update section and added code documentation guidelines
- Release Process - Summary with reference to the detailed instructions in the README
- License - MIT license contribution agreement

nh13 committed:
doc: update contributing guidelines and add issue/PR templates

Rework CONTRIBUTING.md and add GitHub-native issue and pull request
templates.

- Code of Conduct: reference the org-wide Fulcrum Genomics Code of
  Conduct rather than restating a generic statement
- Getting Started: note that the required toolchain is pinned in
  rust-toolchain.toml, so a different toolchain may surface clippy
  lints that CI does not enforce
- Issue Tracker: slim to pointers and add .github/ISSUE_TEMPLATE bug
  report and feature request templates
- Pull Requests: slim to submission steps and add
  .github/pull_request_template.md with a checklist
- Coding Guidelines: reference ./ci/check.sh instead of duplicating
  the fmt/clippy commands
- Code Documentation: add a rustdoc example and a cargo doc --open
  preview tip

clintval commented on CONTRIBUTING.md:
```diff
@@ -1,13 +1,249 @@
+
+## Code of Conduct
+
+We are committed to providing a welcoming and inclusive environment for all contributors. Please be respectful and considerate in all interactions.
```
Aren't we already covered with our global code of conduct?

https://github.com/fulcrumgenomics/.github/blob/main/CODE_OF_CONDUCT.md

nh13 replied:
Good call — updated this section to point at the org-wide [Code of Conduct](https://github.com/fulcrumgenomics/.github/blob/main/CODE_OF_CONDUCT.md) rather than restating a generic statement.

clintval commented on CONTRIBUTING.md:
```diff
@@ -1,13 +1,249 @@
+- Steps to reproduce the issue
+- Expected behavior vs. actual behavior
+- Your environment (OS, Rust version, fqtk version)
+- Any relevant error messages or logs
+- Sample input files if applicable (or a description of the input format)
+
+### Suggesting Features
+
+Feature requests are welcome! When suggesting a feature:
+
+1. Check existing issues to avoid duplicates
+2. Clearly describe the use case and why the feature would be valuable
+3. If possible, outline how you envision the feature working
+
+### Issue Labels
+
+- `bug` - Something isn't working
+- `enhancement` - New feature or request
+- `documentation` - Documentation improvements
+- `good first issue` - Good for newcomers
```
A lot of this could be added to an issue template which is a native GitHub feature:

https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/configuring-issue-templates-for-your-repository

nh13 replied:
Agreed. Added `.github/ISSUE_TEMPLATE/` bug-report and feature-request templates (plus a `config.yml` with a link to the contributing guide) and slimmed this section to a pointer at the templates, keeping the label list.

clintval commented on CONTRIBUTING.md:
```diff
@@ -1,13 +1,249 @@
+- All PRs require review before merging
+- CI checks must pass (formatting, linting, tests)
+- Address reviewer feedback promptly
+- Keep PRs focused and reasonably sized when possible
```
Contributors are probably more likely to read and follow these instructions if we primarily have them in a pull request template:

https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository

nh13 replied:
Done — added `.github/pull_request_template.md` with a checklist and the description / related-issues / breaking-changes prompts, and trimmed this section to the submission steps plus a pointer to the template.

nh13 commented on CONTRIBUTING.md:
```diff
@@ -1,13 +1,249 @@
+### Before Submitting
+
+1. Ensure all tests pass locally: `cargo test`
+2. Run the full pre-commit checks: `./ci/check.sh`
```
The toolchain is pinned to 1.85 in `rust-toolchain.toml` and clippy passes in CI on that version, so the script is stable on the pinned toolchain. Local failures on `main` are almost certainly a newer clippy flagging lints that aren't enforced here. I've called the pinned toolchain out explicitly in Prerequisites so contributors build with it.

nh13 commented on CONTRIBUTING.md:
````diff
@@ -1,13 +1,249 @@
+Run both with:
+```console
+cargo fmt --all
+cargo clippy --all-features --all-targets -- -D warnings
````
Agreed — replaced the duplicated `cargo fmt` / `cargo clippy` block with a pointer to `./ci/check.sh`.

nh13 commented on CONTRIBUTING.md:
````diff
@@ -1,13 +1,249 @@
+When updating the tool's arguments or docstring, refresh the README by running the included script:
 
 ```console
 bash .github/scripts/update-docs.sh
````
Good idea. To keep this PR docs-only I've held off on changing `ci/check.sh`, but I've added a note that CI enforces the README doc-sync in addition to `./ci/check.sh`. Folding the `update-docs.sh` verification into `ci/check.sh` (or adding a `--check`/`--allow-dirty` mode) seems worth doing as a separate change — leaving this thread open to track it.

nh13 replied:
Implemented in 120229f: `ci/check.sh` now has a `Docs` step that builds the debug binary, regenerates the embedded usage via `update-docs.sh`, and fails if the README is left dirty — so running `./ci/check.sh` locally catches README drift the same way CI does. Verified the `update-docs.sh` `sed` works on BSD/macOS sed as well, since this now runs in the macOS pre-commit job too.

nh13 committed:
ci: verify README usage is in sync in ci/check.sh

Fold the README `fqtk demux` usage-sync check into ci/check.sh so that
running it locally catches the same drift CI enforces, rather than only
discovering it on a failed CI run.

The new Docs step builds the debug binary, regenerates the embedded
usage via .github/scripts/update-docs.sh, and fails if the README is
left dirty. Update CONTRIBUTING.md to list the README check among what
ci/check.sh runs and to note it mirrors the CI checks.
